{% extends  "core.html" %}
{% load staticfiles %}

{% block title %}OMA Orthology database: About{% endblock %}
{% block meta_description %}OMA is a method and database for the inference of orthologs among complete genomes. Since 2004 it has been continuesly maintained and improved. We provide browsable orthology predictions, APIs, flat file downloads and a standalone version of the inference algorithm.{% endblock %}

{% block content %}
<style>
    blockquote {

        padding: 10px 20px;
        margin: 0 0 20px;
        font-size: 17.5px;
        border-left: 5px solid #eee;
    }

    html {
  font-size: 1rem;
}

@include media-breakpoint-up(sm) {
  html {
    font-size: 1.2rem;
  }
}

@include media-breakpoint-up(md) {
  html {
    font-size: 1.4rem;
  }
}

@include media-breakpoint-up(lg) {
  html {
    font-size: 1.6rem;
  }
}

</style>
<div class="container mt-5">

    <div class="card text-center bg-light mb-3 mx-auto">
        <div class="card-header">
            <h1 class="display-3 mb-3">OMA</h1>

            <small class="text-muted h3">Web based database interface for orthology prediction
            </small>


            <div class="entete my-4" style="height: 10px;"></div>

            <div class="card bg-light m-3 text-left">

                <div class="card-body">

                    <h4 id='nut' class="panel-heading">
                        OMA in a nutshell
                    </h4>

                    <p>The OMA (“Orthologous MAtrix”) project is a method and database for the inference of orthologs among complete genomes. The distinctive features of OMA are its broad scope and size, <a id='linkToBench' href='#bench'>high quality of inferences</a>, feature-rich web interface, <a href="{% url 'current'  %}">availability of data in a wide range of formats and interfaces</a>, and frequent update schedule of two releases per year.</p>

                    <p>OMA’s inference algorithm consists of three main phases. First, to infer homologous sequences (sequences of common ancestry), all-against-all Smith-Waterman alignments are computed and significant matches are retained. Second, to infer orthologous pairs (the subset of homologs related by speciation events), mutually closest homologs are identified based on evolutionary distances, taking into account distance inference uncertainty and the possibility for differential gene losses (for more details, see <a href="http://www.biomedcentral.com/1471-2105/10/220">Roth et al 2008</a>). Third, these orthologs are clustered in two different ways, which are useful for different purposes: (a) we identify cliques of orthologous pairs (“OMA groups”), which are useful as marker genes for phylogenetic reconstruction and tend to be very specific; (b) we identify hierarchical orthologous groups (“HOGs”), groups of genes defined for particular taxonomic ranges and identify all genes that have descended from a common ancestral gene in that taxonomic range. Fore more details on the algorithm to infer HOGs from orthologous pairs, see <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0053786">Altenhoff et al. 2013</a>.</p>

                    <p>The OMA pipeline can also run on custom genomic/transcriptomic data using the <a href="/standalone/">OMA stand-alone software</a>, and it is even possible to combine precomputed data with custom data by <a href="/export">exporting parts of the OMA database</a>.
                        </p>

                    <p>For more info on the feature of the OMA Browser, please consult the help pages accessible from the navigation bar in the top-right corner.</p>
                    </div>
                </div>

            <div class="card bg-light m-3 text-left">

                <div class="card-body">

                    <h4 id='history' class="panel-heading">History</h4>

                    <p class="lead">{% now 'Y' as current_year %}</p>
                    <p>The OMA project was initiated in 2004 at ETH Zurich by Prof. Gaston Gonnet, with the goal of identifying orthologs among all publicly available genomes. At the time, most sequenced genomes were bacteria and only few were eukaryotes. Several PhD students in his group became increasingly involved, in particular Adrian Schneider, Christophe Dessimoz, and Alexander Roth. Over the subsequent  {{ current_year|add:"-2005" }} years, OMA underwent 20 major releases, steadily increasing the number of genomes under consideration.</p>

                    <img src="{% static "image/history.png" %}" class="img-fluid margin-auto mb-3" alt="Graph showing growth of OMA over time.">


                    <p>The OMA Browser was introduced in 2006. Early releases were developed by Adrian Schneider and Christophe Dessimoz. Adrian Altenhoff joined the team in 2008.
                            </p>


                    <p>In 2008, the responsibility of “baby-sitting” the all-against-all (i.e. importing and converting genomes, running and verifying computations across hundreds of CPUs) was handed over from Gaston to the two Adrians.</p>


                    <p>Since 2010, Adrian Altenhoff has been the main baby-sitter of the all-against-all and manager of OMA’s operations. In 2011, Christophe joined Gaston as co-PI of OMA. In 2012, OMA became a SIB-funded bioinformatics resource.</p>

                </div>

            </div>


            <h3 class="mt-2">Publications</h3>

            <div class="card bg-light m-3 text-left">

                <div class="card-body">

                    <h4 id='publications' class="panel-heading">OMA overview</h4>

                    <p>Adrian M Altenhoff, Natasha M Glover, Clément-Marie Train, Klara Kaleb, Alex Warwick Vesztrocy, David Dylus, Tarcisio Mendes de Farias, Karina Zile, Charles Stevenson, Jiao Long, Henning Redestig, Gaston H Gonnet and Christophe Dessimoz</p>

                    <p><em>The OMA orthology database in 2018: retrieving evolutionary relationships among all domains of life through richer web and programmatic interfaces</em></p>

                    <p>Nucleic Acids Research, 2018, 46 (D1): D477-D485 <a href="http://doi.org/10.1093/nar/gkx1019">Full text</a></p>


                    <p>Adrian M. Altenhoff, Nives Škunca, Natasha Glover, Clément-Marie Train, Anna Sueki, Ivana Piližota, Kevin Gori, Bartlomiej Tomiczek, Steven Müller, Henning Redestig, Gaston H Gonnet and Christophe Dessimoz</p>

                    <p><em>The OMA orthology database in 2015: function predictions, better plant support, synteny view, and other improvements</em></p>
                    <p>Nucleic Acids Research, 2015, 43 (D1): D240-D249 <a href="http://dx.doi.org/10.1093/nar/gku1158"> Full text</a></p>

                    <p>Adrian M. Altenhoff, Adrian Schneider, Gaston H. Gonnet, Christophe Dessimoz</p>

                    <p><em>OMA 2011: Orthology Inference Among 1,000 Complete Genomes</em></p>

                    <p>Nucleic Acids Research, 2011, 39 (suppl 1): D289-D294 <a href='http://nar.oxfordjournals.org/content/39/suppl_1/D289'> Full text</a>
                    </p>

                    <p>Adrian Schneider, Christophe Dessimoz, Gaston H. Gonnet</p>

                    <p><em>OMA Browser - Exploring orthologous relations across 352 complete genomes</em></p>

                    <p>Bioinformatics, 2007, 23:16 (pp. 2180-2182) <a href="/">Project page</a> <a href='http://bioinformatics.oxfordjournals.org/content/23/16/2180'> Full text</a>
                    </p>

                    <div class="entete my-4" style="height: 4px;"></div>

                    <h4 class="panel-heading">OMA algorithm</h4>

                    <p>Clément-Marie Train, Natasha M. Glover, Gaston H. Gonnet, Adrian M. Altenhoff, Christophe Dessimoz</p>

                    <p><em>Orthologous Matrix (OMA) algorithm 2.0: more robust to asymmetric evolutionary rates and more scalable hierarchical orthologous group inference</em></p>

                    <p>Bioinformatics, 2017, 33:14, (pp. i75–i82) <a href="https://doi.org/10.1093/bioinformatics/btx229">Full text</a>
                    </p>

                    <p>Adrian M. Altenhoff, Manuel Gil, Gaston H. Gonnet, Christophe Dessimoz</p>

                    <p><em>Inferring Hierarchical Orthologous Groups From Orthologous Gene Pairs</em></p>

                    <p>PLoS One, 2013, 8:1, e53786 <a href="/standalone/">Software</a> <a href='http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0053786'>Full text</a>
                    </p>

                    <p>Alexander Roth, Gaston H. Gonnet, Christophe Dessimoz</p>

                    <p><em>Algorithm of OMA for Large-Scale Orthology Inference</em></p>

                    <p>BMC Bioinformatics, 2008, 9:518
                        <a href='http://www.biomedcentral.com/1471-2105/9/518'>Full text</a>
                        <a href="http://www.biomedcentral.com/1471-2105/10/220">(Please note the associated erratum)</a>
                    </p>

                    <p>Christophe Dessimoz, Brigitte Boeckmann, Alexander Roth, Gaston H. Gonnet</p>

                    <p><em>Detecting Non-Orthology in the COGs Database and Other Approaches Grouping Orthologs Using Genome-Specific Best Hits</em></p>

                    <p>Nucleic Acids Res, 2006 34:11 (pp. 3309-3316)
                        <a href='http://nar.oxfordjournals.org/content/34/11/3309'>Full text</a>
                    </p>

                    <p>Christophe Dessimoz, Gina Cannarozzi, Manuel Gil, Daniel Margadant, Alexander Roth, Adrian Schneider, Gaston H. Gonnet</p>

                    <p><em>OMA, a Comprehensive, Automated Project for the Identification of Orthologs from Complete Genome Data: Introduction and First Achievements</em></p>

                    <p>RECOMB 2005 Workshop on Comparative Genomics, LNCS 3678 (pp. 61-72) <a href='http://download.springer.com/static/pdf/65/chp%253A10.1007%252F11554714_6.pdf?auth66=1408719784_5261138a05efa9237079a02f2783af34&amp;ext=.pdf'>Full text</a>
                        <a href='http://link.springer.com/chapter/10.1007%2F11554714_6'>Abstract</a>
                    </p>

                    <div class="entete my-4" style="height: 4px;"></div>

                    <h4  class="panel-heading" id='bench'>Benchmarking</h4>

                    <p>Altenhoff AM, Boeckmann B, Capella-Gutierrez S, Dalquen DA, DeLuca T, Forslund K, Huerta-Cepas J, Linard B, Pereira C, Pryszcz LP, Schreiber F, da Silva AS, Szklarczyk D, Train CM, Bork P, Lecompte O, von Mering C, Xenarios I, Sjölander K, Jensen LJ, Martin MJ, Muffato M; Quest for Orthologs consortium, Gabaldón T, Lewis SE, Thomas PD, Sonnhammer E, Dessimoz C.</p>

                    <p><em>Standardized benchmarking in the quest for orthologs</em></p>

                    <p>Nat Methods. 2016 May;13(5):425-30. doi: 10.1038/nmeth.3830 <a href="http://dx.doi.org/10.1038/nmeth.3830">Full text</a>
                    </p>

                    <p >Daniel A. Dalquen, Adrian M. Altenhoff, Gaston H. Gonnet, Christophe Dessimoz</p>

                    <p><em>The Impact of Gene Duplication, Insertion, Deletion, Lateral Gene Transfer and Sequencing Error on Orthology Inference: a Simulation Study</em></p>

                    <p>PLoS One, 2013, 8:2, e56925 <a href='http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0056925'>Full text</a>
                    </p>

                    <p>Brigitte Boeckmann, Marc Robinson-Rechavi, Ioannis Xenarios, Christophe Dessimoz</p>
                    <p><em>Conceptual Framework and Pilot Study to Benchmark Phylogenomic Databases Based on Reference Gene Trees</em></p>

                    <p>Briefings in Bioinformatics, 2011, 12:5 (pp. 474-484) <a href='http://bib.oxfordjournals.org/content/12/5/423'> Full text</a>
                    </p>

                    <p >Adrian M. Altenhoff and Christophe Dessimoz</p>

                    <p><em>Phylogenetic and Functional Assessment of Orthologs Inference Projects and Methods</em></p>

                    <p>PLoS Computational Biology, 2009, 5:1, e1000262
                        <a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1000262'>Full text</a>
                    </p>

                    <div class="entete my-4" style="height: 4px;"></div>

                    <h4 class="panel-heading">Orthology inference in general</h4>

                    <p >Adrian M. Altenhoff and Christophe Dessimoz</p>

                    <p><em>Inferring Orthology and Paralogy Evolutionary Genomics: Statistical and Computational methods</em></p>

                    <p>(M Anisimova, Editor), Methods in Molecular Biology, 2012, Springer Humana, Vol. 855. <a href='http://link.springer.com/content/pdf/10.1007%2F978-1-61779-582-4_9.pdf'>Full text</a>
                        <a href='http://link.springer.com/protocol/10.1007%2F978-1-61779-582-4_9'>Abstract</a>
                    </p>

                </div>
            </div>
        </div>
    </div>
</div>
{% endblock content %}
