{% extends  "base_template.html" %}
{% load staticfiles %}

{% block title %}OMA orthology database: orthologs between two genomes{% endblock %}
{% block meta_description %}Download the list of predicted orthologs between a pair of genomes.{% endblock %}
{% block css_hook %}
    <link type="text/css" rel="stylesheet" href="{% static 'css/typeaheadjs.css' %}"/>
{% endblock css_hook %}
{% block js_hook %}
    <script src="{% static 'js/typeahead.bundle.js' %}"></script>
{% endblock js_hook %}

{% block content %}

<div id="oma-generic-container">

    <div class="container">

        <div class="jumbotron">
            <h1>Genome Pair View

               <span> <small class="text-center text-muted h3 d-block">Download orthologs between two species
                </small></span></h1>

            <div class="oma-seperator">
                <span class="colour1"></span>
                <span class="colour2"></span>
                <span class="colour3"></span>
            </div>

             <div>

            <p class="text-justify">Use the following form to download the list of all predicted orthologs
                        between a pair of genomes of interest. Since orthologs are sometimes 1:many
                        or many:many relations, this download will return more orthologs than what
                        is covered by the OMA groups. The result is returned as a tab-separated
                        text file, each line corresponding to one orthologous relation. The columns
                        are the two IDs, the type of orthology (1:1, 1:n, m:1 or m:n) and (if available)
                        the OMA group containing both sequences.
                    </p>

                 <br>




                 <div class="card card-header-oma">
                <div class="card-header">
                    <h4>Download orthologs between two species</h4>
                </div>
                <div class="card-body">
                     <form class="form " role="form" id="genomePW" action="/cgi-bin/gateway.pl" method="get" >
                    <INPUT TYPE="HIDDEN" NAME="f" VALUE="PairwiseOrthologs">

                        <div class="form-group row">
                            <label for="textG1" class="col-md-2 offset-md-1 control-label">Species 1:</label>
                            <div class="col-sm-7">
                                <input  id="textG1" type="text"  class="form-control typeahead" name="p1">
                            </div>
                            <br>
                        </div>
                        <div class="form-group row">
                            <label for="textG2" class="col-md-2 offset-md-1 control-label">Species 2:
                            </label>
                            <div class="col-sm-7">
                                <input  id="textG2" type="text"  class="form-control typeahead" name="p2">
                            </div>
                            <br>
                        </div>
                        <div class="form-group row">
                            <label for="textG2" class="col-md-2 offset-md-1  control-label">Preferred ID:
                            </label>
                            <div class=" col-xs-6 col-sm-2 ">
                                <select id='selectBrowser' class="form-control " for="inputSuccess3" name="p3">
                                <OPTION value="OMA">OMA IDs</OPTION>
                                <OPTION value="Source">Source Data IDs</OPTION>
                                <OPTION value="SourceAC">Source Data ACs</OPTION>
                                <OPTION value="SwissProt">UniProt/SwissProt IDs</OPTION>
                                <OPTION value="SwissProt_AC">UniProt/SwissProt ACs</OPTION>
                                <OPTION value="UniProt">UniProt ACs</OPTION>
                                <OPTION value="Refseq_AC">RefSeq ACs</OPTION>
                                <OPTION value="EntrezGene">Entrez Gene IDs</OPTION>
                                <OPTION value="EnsemblGene">Ensembl Gene IDs</OPTION>
                                </select>

                                </div>



                        </div>

                         <button type="submit" class="btn btn-lg btn-block btn-primary ">Get pairs</button>

                    </form>
                </div>
            </div>







        </div>
        </div>



    </div>
</div>

<script type="text/javascript">
$(document).ready(function() {

  var genomeNames = new Array();
  var genomeIDs = new Object();
  $.getJSON( "/All/flatgenomes.json", function( jsonData ) {
     $.each( jsonData, function ( id, genome ) {
         var tmp = genome.name+ ' (' + genome.id + ')';
         genomeNames.push ( tmp );
         genomeIDs[tmp] = genome.id;
     });
  });

  var substringMatcher = function(strs) {
    return function findMatches(q, cb) {
      var matches, substrRegex;

      // an array that will be populated with substring matches
      matches = [];
      if (q.length < 2) {
          return(cb(matches));
      }

      // regex used to determine if a string contains the substring `q`
      substrRegex = new RegExp(q, 'i');

      // iterate through the pool of strings and for any string that
      // contains the substring `q`, add it to the `matches` array
      $.each(strs, function(i, str) {
        if (substrRegex.test(str)) {
          // the typeahead jQuery plugin expects suggestions to a
          // JavaScript object, refer to typeahead docs for more info
          matches.push({ value: str });
        }
      });

      cb(matches);
    };
  };


  $('#textG1').typeahead({
    hint: true
  },
  {
    name: 'genomes',
    minLength: 4,
    source: substringMatcher(genomeNames)
  }).blur(function(){
      if(genomeIDs[$(this).val()] == null) {
        $(this).val('');
      }
  });
  $('#textG2').typeahead({
    hint: true
  },
  {
    name: 'genomes',
    minLength: 4,
    source: substringMatcher(genomeNames)
  }).blur(function(){
      if(genomeIDs[$(this).val()] == null) {
        $(this).val('');
      }
  });

  $('#genomePW').submit(function(){
       $('#textG1').val(genomeIDs[$('#textG1').val()]);
       $('#textG2').val(genomeIDs[$('#textG2').val()]);
       if ($('#textG1').val() == '' || $('#textG2').val() == '') {
       alert('Please define the two genomes.');
       return false;
  }
  });
    $(window).keydown(function(event){
      if(event.keyCode == 13) {
        event.preventDefault();
        return false;
      }
  });
});

</script>






{% endblock content %}
