{% extends  "core.html" %}
{% load staticfiles %}
{% load bootstrap3 %}

{% block title %}OMA: uses{% endblock %}

{% block content %}

<div class="container-fluid">
	<div class="row">
		<div class=" col-sm-12 col-md-10 col-md-offset-1">
			<div class="jumbotron">
				<div class="container">
					<h1 class="text-center">Typical uses<br>
					</h1>
					<div class="well well-sm entete"></div>
					<br>
					<p>OMA browser is a web-interface-database for orthology inference, many uses are possibles according to queries and the inference users need.
						We have summup the typicals uses you can do with oma browser and what kind of result you will get.
					</p>
					<br>

					<div class="row"><div class="panel panel-default">
						<div class="panel-heading">
							<h4>What would you like to do ? 
							</h4>
						</div>
						<div class="panel-body">
							<div class="col-md-4 ">
								<ul class="nav nav-pills nav-stacked">
									<li class="active"><a href="#propagate" role="tab" data-toggle="tab">Propagate functional annotations between two genomes</a>
									</li>

									<li ><a href="#Identify" role="tab" data-toggle="tab">
										Identify all orthologs of a gene in a given set of organisms</a>
									</li>

									<li ><a href="#twg" role="tab" data-toggle="tab">
										Allign two whole genomes</a>
									</li>

									<li ><a href="#Align" role="tab" data-toggle="tab">
										Align protein–protein interaction networks between two genomes</a>
									</li>

									<li ><a href="#vertebrates" role="tab" data-toggle="tab">
										Identify all genes in vertebrates that descended from the ancestral β-hemoglobin</a>
									</li>

									<li ><a href="#orthologs" role="tab" data-toggle="tab">
										Identify orthologs to infer a species tree</a>
									</li>
								</ul>
							</div>

							<div class="col-md-8 " >
								<div class="tab-content">
									<div class="tab-pane active" id="propagate">
										<p><strong>Appropriate orthology type</strong></p>
										<p>Genome pair view, filtered to 1:1 orthologs</p>
										<p><strong>Comments</strong></p>
										<p>According to current models, 1:1 orthologs are likely to have a similar function.</p>
									</div>

									<div class="tab-pane" id="Identify">
										<p><strong>Appropriate orthology type</strong></p>
										<p>Protein-centric view</p>
										<p><strong>Comments</strong></p>
										<p>This includes in-paralogs in the target genomes lineages with respect to the speciation event with the query genome</p>
									</div>

									<div class="tab-pane" id="twg">
										<p><strong>Appropriate orthology type</strong></p>
										<p>Genome pair view</p>
										<p><strong>Comments</strong></p>
										<p>Orthologous genes are anchor points.</p>
									</div>

									<div class="tab-pane" id="Align">
										<p><strong>Appropriate orthology type</strong></p>
										<p>Genome pair view</p>
										<p><strong>Comments</strong></p>
										<p>Orthologs can be used to restrict the alignment search space. Thus, only conserved parts are used as anchors.</p>
									</div>

									<div class="tab-pane" id="vertebrates">
										<p><strong>Appropriate orthology type</strong></p>
										<p>Hierarchical group of human β-hemoglobin with ‘Vertebrata’ as taxonomic range</p>
										<p><strong>Comments</strong></p>
										<p>This also retrieves γ-hemoglobin and ε-hemoglobin, which are thought to have diverged within the vertebrates (31).</p>
									</div>

									<div class="tab-pane" id="orthologs">
										<p><strong>Appropriate orthology type</strong></p>
										<p>OMA groups with high-coverage of the species of interest</p>
										<p><strong>Comments</strong></p>
										<p>Since, barring classification error, all sequence pairs within an OMA group are orthologs, the corresponding gene tree is expected to follow the species tree. </p>
									</div>
								</div>
							</div>
						</div>
					</div>
				</div>
			</div>
		</div>
	</div>
</div>



{% endblock content %}